go/service/reachability.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package service
+
+import (
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+	"github.com/keybase/go-framed-msgpack-rpc/rpc"
+)
+
+type reachabilityHandler struct {
+	*BaseHandler
+	libkb.Contextified
+	reachability *reachability
+}
+
+func newReachabilityHandler(xp rpc.Transporter, g *libkb.GlobalContext, reachability *reachability) *reachabilityHandler {
+	return &reachabilityHandler{
+		BaseHandler:  NewBaseHandler(xp),
+		Contextified: libkb.NewContextified(g),
+		reachability: reachability,
+	}
+}
+
+func (h *reachabilityHandler) ReachabilityChanged(_ context.Context, _ keybase1.Reachability) error {
+	return nil
+}
+
+func (h *reachabilityHandler) StartReachability(_ context.Context) (keybase1.Reachability, error) {
+	return h.reachability.start(), nil
+}
+
+func (h *reachabilityHandler) CheckReachability(_ context.Context) (keybase1.Reachability, error) {
+	return h.reachability.check(), nil
+}
+
+type reachability struct {
+	libkb.Contextified
+	sync.Mutex
+	lastReachability keybase1.Reachability
+	started          bool
+	shutdownCh       chan bool
+}
+
+func newReachability(g *libkb.GlobalContext) *reachability {
+	return &reachability{
+		Contextified: libkb.NewContextified(g),
+		shutdownCh:   make(chan bool),
+	}
+}
+
+func (h *reachability) setReachability(r keybase1.Reachability) {
+	if h.lastReachability.Reachable != r.Reachable {
```

Comment:
I think you might need to either have multiple sync.Mutexes or finer-grained lock in `start()` as this can be called from `start` or `check` and when called by `check`, there is no lock.